pegen: fold f-string constants and fix string concatenation

FFY00 opened the pull request:
This patch folds the f-string constants, like they were previously (1),
and makes string concatenation construct a Constant instead of JoinedStr
(2).

1) `f'hello' f'hello'` will now construct
JoinedStr(values=[Constant(value='hellohello')])

2) `'hello' 'hello'` will now construct Constant(value='hellohello')

Signed-off-by: Filipe Laíns <lains@riseup.net>

pablogsal commented:
Good start! Thanks for helping with the project!


You need to also handle the presence of byte strings and reject all those concatenations. I would recommend to do a first pass over the collection checking if there is a single fstrings. In that case you construct a joined value, otherwise you cancatenate all strings/bytes accordingly. If you find a single bytes object but also some string or fstrings, you need to fail.

Check first all the joining rules in the repr to make sure you familiarise yourself with the expected behaviour of str + str, str + fstring, fstring + str, fstring + fstring, str/ fstr + bytes and bytes + str / fstring

FFY00 commented:
> You need to also handle the presence of byte strings and reject all those concatenations

Aren't they already handled here?

https://github.com/we-like-parsers/cpython/blob/9e05991a7ae2aa3bc6e66eced957a50104f5e791/Parser/pegen.c#L2352-L2372

All cases specified in #150 seem to be handled, no?

```python
import ast

for case in (
    '"hello" "hello"',
    '"hello" f"hello"',
    'f"hello" "hello"',
    'b"hello" "hello"',
    'b"hello" f"hello"',
    '"hello" b"hello"',
    'f"hello" b"hello"',
    'b"hello" b"hello"',
    'f"hello" f"hello"',
):
    try:
        print(f'{case}:', ast.dump(ast.parse(case).body[0].value))
    except Exception as e:
        print(f'{case}: failed ({e})')
```

```
$ ./python possible-concatenations.py
"hello" "hello": Constant(value='hellohello')
"hello" f"hello": JoinedStr(values=[Constant(value='hello'), Constant(value='hello')])
f"hello" "hello": JoinedStr(values=[Constant(value='hello'), Constant(value='hello')])
b"hello" "hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
b"hello" f"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
"hello" b"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
f"hello" b"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
b"hello" b"hello": Constant(value=b'hellohello')
f"hello" f"hello": JoinedStr(values=[Constant(value='hello'), Constant(value='hello')])
```

FFY00 commented:
Though the constant folding does not seem to be working.

pablogsal commented:
> > You need to also handle the presence of byte strings and reject all those concatenations
> 
> Aren't they already handled here?
> 
> https://github.com/we-like-parsers/cpython/blob/9e05991a7ae2aa3bc6e66eced957a50104f5e791/Parser/pegen.c#L2352-L2372
> 
> All cases specified in #150 seem to be handled, no?
> 
> ```python
> import ast
> 
> for case in (
>     '"hello" "hello"',
>     '"hello" f"hello"',
>     'f"hello" "hello"',
>     'b"hello" "hello"',
>     'b"hello" f"hello"',
>     '"hello" b"hello"',
>     'f"hello" b"hello"',
>     'b"hello" b"hello"',
>     'f"hello" f"hello"',
> ):
>     try:
>         print(f'{case}:', ast.dump(ast.parse(case).body[0].value))
>     except Exception as e:
>         print(f'{case}: failed ({e})')
> ```
> 
> ```
> $ ./python possible-concatenations.py
> "hello" "hello": Constant(value='hellohello')
> "hello" f"hello": JoinedStr(values=[Constant(value='hello'), Constant(value='hello')])
> f"hello" "hello": JoinedStr(values=[Constant(value='hello'), Constant(value='hello')])
> b"hello" "hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
> b"hello" f"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
> "hello" b"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
> f"hello" b"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
> b"hello" b"hello": Constant(value=b'hellohello')
> f"hello" f"hello": JoinedStr(values=[Constant(value='hello'), Constant(value='hello')])
> ```

@FFY00 is not that there are not handled, is that they need to return exactly the same AST as in the main branch.

pablogsal commented on Parser/pegen.c:
```diff
@@ -2359,41 +2360,45 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
-            asdl_seq_SET(values, current_pos++, elem);
+            if (prev_elem && prev_elem->kind == Constant_kind) {
+                /* fold consequent constants */
+                PyObject *new_str = PySequence_Concat(prev_elem->v.Constant.value,
```
This is O(n**2). We need a PyUnicode_Writer and a PyBytes_Writer here to do this efficiency

pablogsal replied:
Check the previous implementation in `string_parser.c` for reference.

FFY00 replied:
Ack. Let's bring back the bytes branch again then. It previously did `PyBytes_Concat`, should I move it to `_PyBytesWriter`?

pablogsal replied:
We could:

* Get this to work using specialized APIs like `PyUnicode_ConcatAndDel` and friends.
* Try to use the optimized Bytes and Unicode writers to do this efficiency.

I would recommend doing (1) and optimize later, but `PySecuence_Concat` is too generic unfortunately

FFY00 committed:
pegen: fold f-string constants and fix string concatenation

This patch folds the f-string constants, like they were previously (1),
and makes string concatenation construct a Constant instead of JoinedStr
(2).

1) `f'hello' f'hello'` will now construct
JoinedStr(values=[Constant(value='hellohello')])

2) `'hello' 'hello'` will now construct Constant(value='hellohello')

Signed-off-by: Filipe Laíns <lains@riseup.net>

FFY00 commented:
The last commit changes the byte string concatenation to use `_PyBytesWriter` but I couldn't see any performance difference.

pablogsal commented:
> The last commit changes the byte string concatenation to use `_PyBytesWriter` but I couldn't see any performance difference.

How are you measuring? If you measure a full parse, creating the AST is 80% if the work, do you need to instrument the code?

FFY00 commented:
I measured the time it took to parse a file with a bunch of long byte strings being concatenated. I scaled it up a little bit now and it is actually registering worse performance. But I am not sure how meaningful my testing is. How would you recommend benchmarking this?

pablogsal commented:
> I measured the time it took to parse a file with a bunch of long byte strings being concatenated. I scaled it up a little bit now and it is actually registering worse performance. But I am not sure how meaningful my testing is. How would you recommend benchmarking this?

Use perf counters / time counters at the enter and exit of the function and parse a file with a lot of concatenations. If you measure the entire parsing you are going to get a ton of noise because this operation is very very small compared with everything else.

On the other hand, having this into account, if we measure and is not that slow/different, we can just call PyUnicode_ConcatAndDel and PyBytes_ConcatAndDel and call it a day

FFY00 commented:
I added perf counters to the function itself, when interpreting the file generated by the following script I get the following values for the total time spent in `_PyPegen_concatenate_strings2`.

```python
with open('long-string-concatenation.py', 'w') as f:
    f.write(f'b"{"a" * 1000}"\n' * 25000)
```

### `_PyBytesWriter` (last commit)

Run 1: 2.099500s
Run 2: 2.046978s
Run 3: 2.114400s
Run 4: 2.073067s
Run 5: 2.078464s

### `PyBytes_Concat` (just the first commit)

Run 1: 0.140555s
Run 2: 0.137755s
Run 3: 0.152179s
Run 4: 0.140775s
Run 5: 0.141847s

---

The patch for the perf counters:

```patch
diff --git a/Parser/pegen.c b/Parser/pegen.c
index 59f7f7620c..8bdbe5f535 100644
--- a/Parser/pegen.c
+++ b/Parser/pegen.c
@@ -2337,11 +2337,14 @@ _PyPegen_seq_delete_starred_exprs(Parser *p, asdl_seq *kwargs)
     return new_seq;
 }
 
+long double total_time = 0;
+
 expr_ty
 _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
                              int lineno, int col_offset, int end_lineno,
                              int end_col_offset, PyArena *arena)
 {
+    clock_t start_time = clock();
     Py_ssize_t len = asdl_seq_LEN(strings);
     assert(len > 0);
 
@@ -2394,6 +2397,10 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
         if (_PyArena_AddPyObject(arena, res) < 0) {
             return NULL;
         }
+        long double elapsed_time = (long double)(clock() - start_time);
+        total_time += elapsed_time;
+        printf("_PyPegen_concatenate_strings2 elapsed %Lf seconds (total %Lf seconds)\n",
+               elapsed_time / CLOCKS_PER_SEC, total_time / CLOCKS_PER_SEC);
         return _PyAST_Constant(res, NULL, lineno, col_offset, end_lineno, end_col_offset, p->arena);
     }
 
@@ -2461,6 +2468,11 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
         }
     }
 
+    long double elapsed_time = (long double)(clock() - start_time);
+    total_time += elapsed_time;
+    printf("_PyPegen_concatenate_strings2 elapsed %Lf seconds (total %Lf seconds)\n",
+           elapsed_time / CLOCKS_PER_SEC, total_time / CLOCKS_PER_SEC);
+
     if (!f_string_found) {
         assert(n_elements == 1);
         expr_ty elem = asdl_seq_GET(values, 0);
```

---

With `_PyBytesWriter`, we need to iterate over the asdl sequence first to calculate the size and then again to add the bytes, so these results do actually make sense. I think we should drop the last patch.

pablogsal commented:
This is still a bit surprising, but the data is data. Let's go with `PyBytes_ConcatAndDel ` and friends then. Thanks for the experiment @FFY00 👍

FFY00 commented:
Anyway, the first patch does properly fix the AST generation.

```
$ ./python possible-concatenations.py
"hello" "hello": Constant(value='hellohello')
"hello" f"hello": JoinedStr(values=[Constant(value='hellohello')])
f"hello" "hello": JoinedStr(values=[Constant(value='hellohello')])
b"hello" "hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
b"hello" f"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
"hello" b"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
f"hello" b"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
b"hello" b"hello": Constant(value=b'hellohello')
f"hello" f"hello": JoinedStr(values=[Constant(value='hellohello')])
```
```
$ python possible-concatenations.py
"hello" "hello": Constant(value='hellohello')
"hello" f"hello": JoinedStr(values=[Constant(value='hellohello')])
f"hello" "hello": JoinedStr(values=[Constant(value='hellohello')])
b"hello" "hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
b"hello" f"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
"hello" b"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
f"hello" b"hello": failed (cannot mix bytes and nonbytes literals (<unknown>, line 1))
b"hello" b"hello": Constant(value=b'hellohello')
f"hello" f"hello": JoinedStr(values=[Constant(value='hellohello')])
```
```
$ diff <(./python possible-concatenations.py) <(python possible-concatenations.py)
```

FFY00 commented:
Is there anything else to take into account?

pablogsal commented:
> Is there anything else to take into account?

No, I will review the last version later today and we can land it. Great job!

lysnikolaou reviewed (approved):
First of all, thanks @FFY00 for helping out!

Really great work on this. I was looking into it myself this week, but hadn't started writing code for it, and this seems like a really nice way to do it.

Only one comment: Would it make sense (should we benchmark?) to implement some kind of mechanism, in order to return early in case the `string` seq has only one element, since this is going to be the vast majority of the cases? I know this is maybe a premature optimization, but the number of cases where there's only one string there will occur much more frequently.

FFY00 commented:
Thank you, yes, I think it would make sense, we already have all the necessary information :blush:

lysnikolaou commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,81 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
 
-    asdl_expr_seq* values = _Py_asdl_expr_seq_new(n_elements, p->arena);
-    if (values == NULL) {
+    if (unicode_string_found && !f_string_found && len == 1) {
+        return asdl_seq_GET(strings, 0);
+    }
```
Is there some special handling of f-strings in here that prevents us from just cheking for `len == 1`?

FFY00 replied:
`f'{1}'` produces, or should produce, `JoinedStr(values=[FormattedValue(value=Constant(value=1), conversion=-1)])`, so yes.

Currently, we seem to be producing `JoinedStr(values=[FormattedValue(value=Constant(value=1), conversion=-1), Constant(value='')])`. I am not sure if that should be fixed in grammar, or in this function. If we fix it here, by skipping empty strings, we can just do `len == 1`.

FFY00 replied:
But as-is we should actually be able to drop the `unicode_string_found &&`.

lysnikolaou replied:
I'm fixing this exact problem in #164.

FFY00 replied:
:rocket:

isidentical commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,81 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+        return asdl_seq_GET(strings, 0);
+    }
+
+    asdl_expr_seq* flatenned = _Py_asdl_expr_seq_new(n_flattened_elements, p->arena);
```
small nit: `flatenned` -> `flattened`

isidentical commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,81 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+                }
+            }
+            i = j-1;
+            expr_ty concat_str = _PyAST_Constant(_PyUnicodeWriter_Finish(&writer),
+                                                 NULL, lineno, col_offset,
+                                                 end_lineno, end_col_offset, p->arena);
```
Question: what does these positions belong to? Wouldn't it be more precesine if we do `first_elem->lineno, first_elem->col_offset, last_elem->end_lineno, last_elem->end_col_offset`?

FFY00 replied:
Hum, yes.

FFY00 replied:
Can you check if it looks alright now?

isidentical reviewed (approved):
Thanks! LGTM

FFY00 committed:
pegen: optimize _PyPegen_concatenate_strings2 for single strings

Signed-off-by: Filipe Laíns <lains@riseup.net>

isidentical commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,82 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+                }
             }
+            i = j-1;
+            expr_ty concat_str = _PyAST_Constant(_PyUnicodeWriter_Finish(&writer),
```
I think you should check `_PyUnicodeWriter_Finish(&writer) != NULL` above or return `NULL`

pablogsal replied:
`concat_str` can be NULL

isidentical commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,82 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+            for (j = i; j < n_flattened_elements; j++) {
+                elem2 = asdl_seq_GET(flattened, j);
+                if (elem2->kind == Constant_kind) {
+                    _PyUnicodeWriter_WriteStr(&writer, elem2->v.Constant.value);
```
You also need to check this as well, we might get out of memory etc.

FFY00 replied:
Oh, right :facepalm:

pablogsal commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,82 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+            for (j = i; j < n_flattened_elements; j++) {
+                elem2 = asdl_seq_GET(flattened, j);
+                if (elem2->kind == Constant_kind) {
+                    _PyUnicodeWriter_WriteStr(&writer, elem2->v.Constant.value);
```
Check the error return code of `_PyUnicodeWriter_WriteStr` and react accordingly

isidentical commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,82 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+            i = j-1;
+            expr_ty concat_str = _PyAST_Constant(_PyUnicodeWriter_Finish(&writer),
+                                                 NULL, elem->lineno, elem->col_offset,
+                                                 elem2->end_lineno, elem2->end_col_offset, p->arena);
```
Is `elem2` the last `Constant` element or the last non-Constant element. If it is the latter case, I think in the `else {break}` we also have to set `elem2` to the previous item in the sequence. Or just use `asdl_seq_GET(flattened, j-1)`

FFY00 replied:
Oh, right :facepalm:

pablogsal commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,82 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+            expr_ty concat_str = _PyAST_Constant(_PyUnicodeWriter_Finish(&writer),
+                                                 NULL, elem->lineno, elem->col_offset,
+                                                 elem2->end_lineno, elem2->end_col_offset, p->arena);
+            asdl_seq_SET(values, current_pos++, concat_str);
+        } else {
+            asdl_seq_SET(values, current_pos++, elem);
```
You can probably join these and remove the else by doing `elem = concat_str` in the first branch or similar

pablogsal commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,82 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+            i = j-1;
+            expr_ty concat_str = _PyAST_Constant(_PyUnicodeWriter_Finish(&writer),
+                                                 NULL, elem->lineno, elem->col_offset,
+                                                 elem2->end_lineno, elem2->end_col_offset, p->arena);
```
`elem2` here can be used uninitialized as the code stands currently. You should get a compiler warning.

FFY00 replied:
The loop should run at least one time for `j = i`, both loops have the same check (`i < n_flattened_elements` and `j < n_flattened_elements`), but I do seem to be getting the compiler warning, I missed that somehow earlier. I will initialize it anyway.

pablogsal commented:
I noticed an issue with new lines while testing this PR:

https://github.com/we-like-parsers/cpython/issues/166

pablogsal commented:
@isidentical Haha, we made the same review :P

FFY00 commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,90 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+    for (i = 0; i < n_flattened_elements; i++) {
+        expr_ty elem = asdl_seq_GET(flattened, i);
+        if (elem->kind == Constant_kind) {
+            _PyUnicodeWriter_Init(&writer);
+            expr_ty last_elem = elem;
+            for (j = i; j < n_flattened_elements; j++) {
+                elem = asdl_seq_GET(flattened, j);
+                if (elem->kind == Constant_kind) {
+                    if (_PyUnicodeWriter_WriteStr(&writer, elem->v.Constant.value)) {
+                        return NULL;
+                    }
+                    last_elem = elem;
+                } else {
+                    break;
+                }
+            }
```
There is an optimization to be done here, we could skip `_PyUnicodeWriter` and creating a new AST constant altogether if it's just one string. I did not do it initially for simplicity's sake, should I do it now?

pablogsal replied:
Yeah, it makes sense because that's like the dominant path for f-strings.

pablogsal replied:
But be careful, we are already burning a lot of indentation levels and this code is already difficult to read

pablogsal replied:
Ideally you can do an early return close to the exit for 1 string that is not an f-string before:

```
    if (!f_string_found && len == 1) {
        return asdl_seq_GET(strings, 0);
    }
```

FFY00 replied:
I already do that optimization, see the 2nd commit.

> But be careful, we are already burning a lot of indentation levels and this code is already difficult to read

With my refactor, this is now actually much easier to do, I can just add some clausules to the if that decides if we for for the `_PyUnicodeWriter` branch.

FFY00 replied:
I added a comment and some spacing to make the code easier to digest.

pablogsal replied:
> I already do that optimization, see the 2nd commit.

I know you do, I am suggesting to do the one you are describing now close to that one

FFY00 replied:
Ah, yes, sorry. I just needed to add `i+1 < n_flattened_elements && asdl_seq_GET(flattened, i+1)->kind == Constant_kind` to the `if`.

pablogsal commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,90 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+                elem = asdl_seq_GET(flattened, j);
+                if (elem->kind == Constant_kind) {
+                    if (_PyUnicodeWriter_WriteStr(&writer, elem->v.Constant.value)) {
+                        return NULL;
```
You need to deallocate the writer in this branch: `_PyUnicodeWriter_Dealloc`

pablogsal replied:
Not sure if you forgot to commit or you are still on it, but just in case: You are still missing this.

pablogsal replied:
`_PyUnicodeWriter_Dealloc` needs to be called before returning `NULL` and only in the error paths

pablogsal commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,99 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+            i = j-1;
+
+            PyObject *concat_str = _PyUnicodeWriter_Finish(&writer);
+            _PyUnicodeWriter_Dealloc(&writer);
```
No, this is incorrect. This needs to be done in the error path only:

https://github.com/we-like-parsers/cpython/pull/149/files#r725655273

FFY00 replied:
Ah :sweat_smile:

FFY00 commented:
Rebased to solve conflicts as #164 touched some whitespace :stuck_out_tongue:

pablogsal commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,100 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+    int prev_is_constant = 0;
+    for (i = 0; i < n_flattened_elements; i++) {
+        expr_ty elem = asdl_seq_GET(flattened, i);
+        if (!(prev_is_constant && elem->kind == Constant_kind)) {
```
```suggestion
        if (!prev_is_constant || elem->kind != Constant_kind)) {
```

pablogsal committed:
Simplify condition

pablogsal commented on Parser/pegen.c:
```diff
@@ -2383,28 +2383,100 @@ _PyPegen_concatenate_strings2(Parser *p, asdl_expr_seq *strings,
+    int prev_is_constant = 0;
+    for (i = 0; i < n_flattened_elements; i++) {
+        expr_ty elem = asdl_seq_GET(flattened, i);
+        if (!prev_is_constant || elem->kind != Constant_kind)) {
```
```suggestion
        if (!prev_is_constant || elem->kind != Constant_kind) {
```

pablogsal commented:
Grteat work @FFY00! :tada: 

I am merging this so we can start building on top.
